p2p/p2p-core/src/transports/tcp.rs

```diff
@@ -0,0 +1,156 @@
+//!
+
+use std::{
+    net::{Ipv4Addr, Ipv6Addr, SocketAddr},
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use crate::{NetworkZone, Transport};
+use cuprate_wire::MoneroWireCodec;
+use futures::Stream;
+use tokio::net::{
+    tcp::{OwnedReadHalf, OwnedWriteHalf},
+    TcpListener, TcpStream,
+};
+use tokio_util::codec::{FramedRead, FramedWrite};
+
+/// Classic, TCP Socket based default transport.
+#[derive(Debug, Clone, Copy, Default)]
+pub struct Tcp;
+
+#[derive(Debug, Clone)]
+/// Mandatory parameters for starting the TCP p2p inbound listener
+pub struct TcpServerConfig {
+    /// Listening IPv4 Address.
+    pub ipv4: Option<Ipv4Addr>,
+    /// Listening IPv6 Address.
+    pub ipv6: Option<Ipv6Addr>,
+
+    /// Listening IPv4 Port.
+    pub port: u16,
+
+    /// Number of milliseconds before timeout at TCP writing
+    _send_timeout: Duration,
+}
+
+impl Default for TcpServerConfig {
+    fn default() -> Self {
+        Self {
+            ipv4: Some(Ipv4Addr::UNSPECIFIED),
+            ipv6: None,
+            port: 18081,
+            _send_timeout: Duration::from_secs(20),
+        }
+    }
+}
+
+/// A set of listener to which new peers can connect to
+pub struct TcpInBoundStream {
+    /// IPv4 TCP listener
+    listener_v4: Option<TcpListener>,
+    /// IPv6 TCP listener
+    listener_v6: Option<TcpListener>,
+}
+
+impl Stream for TcpInBoundStream {
+    type Item = Result<
+        (
+            Option<SocketAddr>,
+            FramedRead<OwnedReadHalf, MoneroWireCodec>,
+            FramedWrite<OwnedWriteHalf, MoneroWireCodec>,
+        ),
+        std::io::Error,
+    >;
+
+    /// SAFETY: Caller must ensure that at least one of the listener is `Some`, otherwise this function
+    /// will always return `Poll::Pending`
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.listener_v4
+            .as_ref()
+            .and_then(|l| match l.poll_accept(cx) {
+                Poll::Ready(r) => Some(Poll::Ready(r)),
+                Poll::Pending => None,
+            })
+            .or(self
+                .listener_v6
+                .as_ref()
+                .and_then(|l| match l.poll_accept(cx) {
+                    Poll::Ready(r) => Some(Poll::Ready(r)),
+                    Poll::Pending => None,
+                }))
+            .unwrap_or(Poll::Pending)
+            .map_ok(|(stream, mut addr)| {
+                let ip = addr.ip().to_canonical();
+                addr.set_ip(ip);
+
+                let (read, write) = stream.into_split();
+                (
+                    Some(addr),
+                    FramedRead::new(read, MoneroWireCodec::default()),
+                    FramedWrite::new(write, MoneroWireCodec::default()),
+                )
+            })
+            .map(Some)
+    }
+}
+
+#[async_trait::async_trait]
+impl<Z: NetworkZone<Addr = SocketAddr>> Transport<Z> for Tcp {
+    type ClientConfig = Option<()>;
```

Comment:
```suggestion
    type ClientConfig = ();
```

I don't like this being an option ngl